app/src/main/java/br/com/alexf/boraprofut/features/drawTeams/useCases/TeamDrawerUseCase.kt

```diff
@@ -1,48 +1,79 @@
 package br.com.alexf.boraprofut.features.drawTeams.useCases
 
 import br.com.alexf.boraprofut.models.Player
-import kotlin.math.ceil
 
 class TeamDrawerUseCase {
 
     fun drawRandomTeams(
         players: Set<Player>,
         playersPerTeam: Int
     ): List<Set<Player>> {
-        return players
-            .shuffled()
-            .chunked(playersPerTeam) {
-                it.toSet()
-            }
+        return getTeams(players = players, playersPerTeam = playersPerTeam)
     }
 
     fun drawBalancedTeams(
         players: Set<Player>,
         playersPerTeam: Int
     ): List<Set<Player>> {
-        val amountTeams = ceil(players.size / playersPerTeam.toFloat()).toInt()
-        val sortedPlayers = players.shuffled()
-            .sortedBy {
-                it.level
-            }.toMutableList()
-
-        val teamsWithBestPlayersIncluded = MutableList(amountTeams) {
-            val bestPlayer = sortedPlayers.removeLast()
-            mutableListOf(bestPlayer)
+        return getTeams(isBalancedTeam = true, players = players, playersPerTeam = playersPerTeam)
+    }
+
+    private fun getTeams(
+        isBalancedTeam: Boolean = false,
+        players: Set<Player>,
+        playersPerTeam: Int
+    ): List<Set<Player>> {
+        val onlyGoalKeepers = players
+            .filter { it.isGoalKeeper }
+            .shuffled()
+            .toMutableList()
+        val onlyPlayers = if (isBalancedTeam) {
+            players.filter { !it.isGoalKeeper }.sortedByDescending { it.level }.toMutableList()
+        } else {
+            players.filter { !it.isGoalKeeper }.shuffled().toMutableList()
         }
+        val totalOfGoalKeepers = onlyGoalKeepers.size
+        val totalOfTeams: Int = players.size / playersPerTeam
+        val teams: MutableList<MutableList<Player>> =
+            MutableList(totalOfTeams) { mutableListOf() }
+        var firstTeamWithoutGoalKeeper = 0
```

Comment:
declare a variável apenas quando for usar. Se for o caso, extraia esse código para uma função responsável apenas para criar os times sem goleiro

Reply:
A variável firstTeamWithoutGoalKeeper é utilizado no momento da criação dos times para saber qual vai ser o primeiro time sem goleiro e então adicionar um jogador para completar o time.